### PR TITLE
Book a Reservation - Backend Implementation

### PR DESCRIPTION
**Pull request by jaimevillegas**
Hi there!

@OsvaldoBC, @johnie261, @blueberry1312, and I have made the following changes to the repo:

- Configure a new RoR application with PostgreSQL database.
- Create a database, create tables, models and controllers.
- Fill the database with sample data and seed.
- Create API endpoints to bring the frontend team the ability to manipulate the data.
- Make API Documentation using rswag.
- Make sure that there are no linter errors.

Here you can see the PR for the frontend repo:
https://github.com/jaimevillegas/book-a-reservation-frontend/pull/11

Thank you for taking the time to review our code! ⚡ 🚀

### CODE REVIEW

File: app/controllers/reservation_controller.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+class ReservationController < ApplicationController
+  before_action :read_reservation, only: [:destroy]
+  skip_before_action :verify_authenticity_token, only: [:create]
+
+  def index; end
+
+  def list_reservations
+    @user_name = User.pluck(:name)
+    @reservation = Reservation.joins(:user).pluck(:reservation_date, :due_date, :service_fee, 'users.name',
```
Comment (mcrd25) [inline app/controllers/reservation_controller.rb:9]: I see no error handlign here

File: app/controllers/reservation_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      render json: @reservation
+    else
+      render json: { error: @reservation.errors.full_messages }
+    end
+  end
+
+  private
+
+  def read_reservation
+    @reservation = Reservation.find(params[:id])
```
Comment (mcrd25) [inline app/controllers/reservation_controller.rb:27]: I see no error handlign here

File: app/controllers/glamping_controller.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  def index; end
+
+  def list_glampings
+    @glampings_name = Glamping.pluck(:id, :name, :glamping_type, :image)
+    render json: @glampings_name
+  end
+
+  def list_glampings_details
+    @item = Glamping.find_by(id: params[:id])
+    render json: @item
```
Comment (mcrd25) [inline app/controllers/glamping_controller.rb:13]: I see no error handlign here

File: app/controllers/glamping_controller.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def destroy_glamping
+    @item = Glamping.find_by(id: params[:id])
+    @item.destroy
```
Comment (mcrd25) [inline app/controllers/glamping_controller.rb:18]: I see no error handlign here

File: app/controllers/glamping_controller.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
+class GlampingController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: %i[destroy_glamping create]
+
+  def index; end
+
+  def list_glampings
+    @glampings_name = Glamping.pluck(:id, :name, :glamping_type, :image)
+    render json: @glampings_name
```
Comment (mcrd25) [inline app/controllers/glamping_controller.rb:8]: I see no error handlign here

File: app/controllers/glamping_controller.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
+class GlampingController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: %i[destroy_glamping create]
+
+  def index; end
```
Comment (mcrd25) [inline app/controllers/glamping_controller.rb:4]: If yo uare not using this, it is best to remove it

File: app/controllers/user_controller.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class UserController < ApplicationController
+  def index; end
+
+  def list_users
+    @users_name = User.pluck(:id, :name)
+    render json: @users_name
```
Comment (mcrd25) [inline app/controllers/user_controller.rb:6]: I see no error handlign here

File: app/models/reservation.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class Reservation < ApplicationRecord
+  belongs_to :user
+  belongs_to :glamping
+
+  validates :glamping_id, presence: true
+  validates :user_id, presence: true
+  validates :reservation_date, presence: true
+  validates :due_date, presence: true
```
Comment (mcrd25) [inline app/models/reservation.rb:8]: [OPTIONAL] IS there any other validation yo ucan think of? should due date be only in a certain range? Like not before the day of reservation?

**Review by mcrd25 — CHANGES_REQUESTED** [review]
Hi Team,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Project has test 👍 
- You have API documentation with swagger 👍 
- You have met the functional requirements of the project for your group number 👍 

## Required Changes ♻️
_Check the comments under the review._
- You are missing unit tests 
- 
## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [OPTIONAL] removing the `test` directory since you are using `rspec` for testing and not native testing.
- [OPTIONAL] Consider using the `timecop` gem to freeze time when running tests. It can help avoid errors: https://github.com/travisjeffery/timecop
[OPTIONAL] Should always have positive and negative tests to ensure it works the way you expect it. Eg. What do you expect the output to be when it's not valid or when valid etc.
Consider making all your tests more extensive, not just this one. For this simple application, it may not catch a possible crippling bug, but in bigger applications, it definitely can. Consider this for all projects moving forward.


**Hope all is well, and that my comments were helpful. 😸**

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by jaimevillegas** [comment]
Hello, @mcrd25!

Thank you so much for reviewing our code!

I have a question. Implementing tests is not part of the requirements of the projects. Do we need to add them, or is it an optional feature?

Thank you again! ⚡

**Pull request merged**